### PR TITLE
feat(proto)!: grpc server and client support

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

We need support for gRPC in the rs-tenderdash-abci library

## What was done?

Added feature to generate gRPC code in tenderdash-proto.

In case of tenderdash-proto usage, tenderdash-abci is not required.

## How Has This Been Tested?

Integrated into Platform's rs-drive-abci as part of multi-thread effort.

Added new feature `grpc` to tenderdash-abci and tenderdash-proto.

## Breaking Changes

no_std support in tenderdash-proto requires default-features=false and is experimental

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: abci/tests/grpc.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
+//! Test gRPC server for ABCI protocol.
+//!
+//! This test verifies that the gRPC server generated with tonic as part of the
+//! tenderdash-proto crate can successfully connect to Tenderdash instance.
+//!
+//! This test should be implemented in the tenderdash-proto crate; however, it
+//! is implemented here to use already existing docker container testing
+//! logic.
+#![cfg(feature = "grpc")]
```
Comment (shumkov) [inline abci/tests/grpc.rs:9]: Is this for Arc?

File: proto-compiler/src/lib.rs [hunk 2]
```diff
@@ -107,6 +107,14 @@ pub fn proto_compile() {
...
 
     println!("[info] => Creating structs.");
+
+    #[cfg(feature = "grpc")]
+    tonic_build::configure()
+        .generate_default_stubs(true)
+        .compile_with_config(pb, &protos, &proto_includes_paths)
+        .unwrap();
+
+    #[cfg(not(feature = "grpc"))]
```
Comment (shumkov) [inline proto-compiler/src/lib.rs:117]: If you need to both gRPC and socket server?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: abci/tests/grpc.rs (hunk 1 above)
Comment (lklimek) [reply abci/tests/grpc.rs:9]: This is for whole file.

> Inner attributes, written with a bang (!) after the hash (#), apply to the item that the attribute is declared within. Outer attributes, written without the bang after the hash, apply to the thing that follows the attribute.

https://doc.rust-lang.org/reference/attributes.html

File: proto-compiler/src/lib.rs (hunk 2 above)
Comment (lklimek) [reply proto-compiler/src/lib.rs:117]: For grpc, we compile using tonic (with client/server). If no grpc, we compile directly using prost (without client/server)

File: proto/src/protobuf.rs [hunk 3]
```diff
@@ -54,3 +62,49 @@ pub struct Duration {
     #[prost(int32, tag = "2")]
     pub nanos: i32,
 }
+
+impl serde::Serialize for Duration {
```
[the remark this answers is not in the document]
Comment (lklimek) [reply proto/src/protobuf.rs]: extracted to #54

**Review by shumkov — APPROVED** [review]
👍

**Pull request merged**